filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,331 @@
+
+		require.Nil(t, ctx.Err())
+		rg.remove("test-id")
+
+		require.Equal(t, 0, len(rg.table))
+		require.Error(t, ctx.Err(), context.Canceled)
+
+		newCtx, newCf, err := rg.newContext("test-id", context.Background())
+		requireGroupSuccess(t, newCtx, newCf, err)
+		require.Equal(t, 1, len(rg.table))
+		require.Nil(t, newCtx.Err())
+	})
+}
+
+func TestDefaultHarvesterGroup(t *testing.T) {
+	source := &testSource{"/path/to/test"}
+
+	t.Run("assert a harvester is started in a goroutine", func(t *testing.T) {
+		var wg sync.WaitGroup
+		mockHarvester := &mockHarvester{onRun: correctOnRun, wg: &wg}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		gorountineChecker := newGoroutineChecker()
+
+		wg.Add(1)
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		// wait until harvester.Run is done
+		wg.Wait()
+		// wait until goroutine that started `harvester.Run` is finished
+		gorountineChecker.waitOriginalCount()
+
+		require.Equal(t, 1, mockHarvester.runCount)
+
+		// when finished removed from bookeeper
+		_, ok := hg.readers.table[source.Name()]
+		require.False(t, ok)
+		// stopped source can be stopped
+		require.Nil(t, hg.StopGroup())
+
+		gorountineChecker.waitOriginalCount()
+	})
+
+	t.Run("assert a harvester can be stopped and removed from bookkeeper", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		// run commands while harvester is running
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			// wait until harvester is started
+			if mockHarvester.runCount == 1 {
+				// assert that it is part of the bookkeeper before it is stopped
+				_, ok := hg.readers.table[source.Name()]
+				require.True(t, ok)
+				// after started, stop it
+				hg.Stop(source)
+				// wait until the harvester is stopped
+				gorountineChecker.waitOriginalCount()
+			}
+		})
+
+		// when finished removed from bookeeper
+		_, ok := hg.readers.table[source.Name()]
+		require.False(t, ok)
+	})
+
+	t.Run("assert a harvester for same source cannot be started", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(inputCtx, source)
+		hg.Start(inputCtx, source)
+
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			// error is expected as a harvester group was expected to start twice for the same source
+			err := hg.StopGroup()
+			require.Error(t, err)
+		})
+
+		require.Equal(t, 1, mockHarvester.runCount)
+	})
+
+	t.Run("assert a harvester panic is handled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: panicOnRun}
```

Comment:
`panivOnRun` throws an error value. What if the value is a string?

Reply:
Same behaviour is expected.